### PR TITLE
Worker improvements

### PR DESCRIPTION
**Pull request by MatthewARoy**
## Pull request summary

Upgrade now related to controller level.
Minor downgrade/attack tweaks.
Extraction directive removed for damaged room.

## Testing checklist:

- [x] Changes are backward-compatible OR version migration code is included
- [x] Codebase compiles with current `tsconfig` configuration
- [x] Tested changes on *{choose PUBLIC/PRIVATE}* server OR changes are trivial (e.g. typos)

### CODE REVIEW

File: src/directives/resource/extract.ts [hunk 1]
```diff
@@ -47,6 +47,9 @@ export class DirectiveExtract extends Directive {
 		if (this.colony.level < 6) {
 			log.notify(`Removing extraction directive in ${this.pos.roomName}: room RCL insufficient.`);
 			this.remove();
+		}  else if (!this.colony.terminal) {
```
Comment (bencbartlett) [inline src/directives/resource/extract.ts:50]: good catch; will add condition in `Colony.ts` to not place directive unless terminal is present

File: src/overlords/core/worker.ts [hunk 2]
```diff
@@ -334,9 +334,16 @@ export class WorkerOverlord extends Overlord {
...
-			// Upgrade controller if close to downgrade
-			if (this.colony.controller.ticksToDowngrade <= (this.colony.level >= 4 ? 10000 : 2000)) {
+
+			// TODO Add high priority to block controller with ramparts/walls in case of downgrade attack
+			// FIXME workers get stalled at controller in case of downgrade attack
+			// Upgrade controller if close to downgrade or if getting controller attacked/was downgraded
+			const downgradeLevel = CONTROLLER_DOWNGRADE[this.colony.controller.level] * .7;
+			if ((!this.colony.controller.upgradeBlocked || this.colony.controller.upgradeBlocked < 30)
+				&& (this.colony.controller.ticksToDowngrade <= downgradeLevel
+					|| this.colony.controller.progress > this.colony.controller.progressTotal)) {
```
Comment (bencbartlett) [inline src/overlords/core/worker.ts:345]: seems like setting downgrade threshold to `CONTROLLER_DOWNGRADE[this.colony.controller.level] * .7` would make workers overly busy upgrading controllers at low RCL

**Review by bencbartlett — COMMENTED** [review]
good changes but the worker downgrade change looks like it could have adverse effects at low RCL, so I don't want to merge before botarena

File: src/directives/resource/extract.ts (hunk 1 above)
Comment (MatthewARoy) [reply src/directives/resource/extract.ts:50]: Added

File: src/overlords/core/worker.ts (hunk 2 above)
Comment (MatthewARoy) [reply src/overlords/core/worker.ts:345]: Fixed

**Pull request closed without merging**